### DISCUSSION TITLE
Can I use force:constraints: to make Renovate use a specific version of Poetry?

### DISCUSSION THREAD

**Discussion by grzenkom** [renovatebot/renovate#26309] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.98-full

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### What is your question?

The [documentation](https://docs.renovatebot.com/configuration-options/#constraints) lists `poetry` under `supportedManagers`.

However, the following configuration seems to have no effect (i.e. Renovate still updates `poetry.lock` with the latest version of Poetry):

```json
"force": {
  "constraints": {
    "poetry": "1.2.0"
  }
}
```

Is it possible to use the setting [`constraints`](https://docs.renovatebot.com/configuration-options/#constraints) to enforce a specific version of Poetry to be used?

---

I checked the code and I cannot see `constraints` read anywhere to determine Poetry version:

https://github.com/renovatebot/renovate/blob/dab6a9c908f1dd31a4fa0580939ba41f79d546fc/lib/modules/manager/poetry/artifacts.ts#L50-L79

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment, marked as answer]
You are using the `-full` image of Renovate, which:
- is prebuilt with each manager preinstalled, usually the latest
- does _not_ dynamically select the version of managers to install at runtime

To change that, configure binarySource=install

Poetry supports constraints, and does not need `force` either: https://github.com/renovatebot/renovate/blob/6884d83a0573ebd40210cf0717c1814afd68a649/lib/modules/manager/poetry/artifacts.ts#L184-L186

**Reply by grzenkom** [reply]
For reference, the option is [`binarySource`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource):

> Starting in v36, Renovate's default Docker image (previously referred to as the "slim" image) uses `binarySource=install` while the "full" Docker image uses `binarySource=global`.
